File: calculator/src/main/java/com/turlir/calculator/member/Value.java
package com.turlir.calculator.member;

import com.turlir.calculator.converter.Member;
import com.turlir.calculator.converter.Visual;
import com.turlir.calculator.interpreter.NotationInterpreter;

import java.math.BigDecimal;
import java.math.MathContext;

public class Value implements Member {

    private final BigDecimal mValue;
    private final boolean isFloat;

    public Value(double value) {
        mValue = new BigDecimal(value, MathContext.UNLIMITED);
        isFloat = false;
    }

    public Value(String value, boolean contains) {
        mValue = new BigDecimal(value, MathContext.UNLIMITED);
        isFloat = contains;
    }

    @Override
    public boolean operand() {
        return true;
    }

    @Override
    public void process(NotationInterpreter interpreter) {
        interpreter.pushDigit(mValue);
    }

    @Override
    public Visual view() {
        return new ValuePrint(mValue, isFloat);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Value value = (Value) o;

        return value.mValue.compareTo(mValue) == 0;
    }

    @Override
    public int hashCode() {
        long temp = Double.doubleToLongBits(mValue.doubleValue());
        return (int) (temp ^ (temp >>> 32));
    }

}

File: calculator/src/main/kotlin/com/turlir/calculator/member/Value.kt
package com.turlir.calculator.member

import com.turlir.calculator.converter.Member
import com.turlir.calculator.converter.Visual
import com.turlir.calculator.interpreter.NotationInterpreter

import java.math.BigDecimal
import java.math.MathContext

class Value : Member {

    private val mValue: BigDecimal
    private val isFloat: Boolean

    constructor(value: Double) {
        mValue = BigDecimal(value, MathContext.UNLIMITED)
        isFloat = false
    }

    constructor(value: String, contains: Boolean) {
        mValue = BigDecimal(value, MathContext.UNLIMITED)
        isFloat = contains
    }

    override fun operand() = true

    override fun view(): Visual = ValuePrint(mValue, isFloat)

    override fun process(interpreter: NotationInterpreter) {
        interpreter.pushDigit(mValue)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false

        val value = other as Value?

        return value!!.mValue.compareTo(mValue) == 0
    }

    override fun hashCode(): Int {
        val temp = java.lang.Double.doubleToLongBits(mValue.toDouble())
        return (temp xor temp.ushr(32)).toInt()
    }

}
